### PR TITLE
Cache results of entitlement type conversion

### PR DESCRIPTION
**Pull request by dsainati1**
Optimize entitlements migration by caching results of `ConvertToEntitledType`. 

Depends on https://github.com/onflow/cadence/pull/3272

______

- [X] Targeted PR against `master` branch
- [X] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [X] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [X] Updated relevant documentation 
- [X] Re-reviewed `Files changed` in the Github PR explorer
- [X] Added appropriate labels

### CODE REVIEW

**Comment by turbolent** [comment]
Nice! Good idea to cache the migration result. Maybe we can lift this from a per-account cache, to a global cache? 🤔

File: runtime/sema/type.go [hunk 1]
```diff
@@ -693,7 +693,7 @@ func (t *OptionalType) QualifiedString() string {
 }
 
 func FormatOptionalTypeID[T ~string](elementTypeID T) T {
-	return T(fmt.Sprintf("%s?", elementTypeID))
+	return T(fmt.Sprintf("(%s)?", elementTypeID))
```
[the remark this answers is not in the document]
Comment (SupunS) [reply runtime/sema/type.go:696]: I think this would impact storage. e.g: dictionary keys with type-values. TypeID is used for hashing of type-value

**Comment by dsainati1** [comment]
Yeah I think I can add a function here to create a migration with a pre-existing cache? And the flow-go code can pass that in.

File: runtime/sema/type.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (turbolent) [reply runtime/sema/type.go:696]: Ugh, good catch! That's a can of worms ... How would we even fix this?

File: migrations/entitlements/migration.go [hunk 2]
```diff
@@ -72,14 +87,30 @@ func ConvertToEntitledType(
...
+
+	if migratedType, exists := migratedTypeCache[staticTypeID]; exists {
+		return migratedType, nil
+	}
+
+	defer func() {
+		if resultType != nil && conversionErr == nil {
+			migratedTypeCache[staticTypeID] = resultType
+		}
+	}()
```
Comment (turbolent) [inline migrations/entitlements/migration.go:103]: The cache might get accessed concurrently (cache is shared between account migrations, accounts are migrated concurrently). Maybe Make this a `sync.Map` and wrap it in a "`StaticTypeCache`"

**Comment by dsainati1** [comment]
@SupunS @turbolent now that https://github.com/onflow/cadence/issues/3248 is merged, this should be unblocked. Could you take another look?

**Comment by turbolent** [comment]
@dsainati1 yeah, we should be able to look into this again. Could you please update the branch?

**Comment by turbolent** [comment]
Not 100% sure, but given that the entitlements migration runs after the static type migration, I guess it should be safe to use the type ID as the cache key? Alternatively we could maybe also use the encoding (CBOR bytes) of the static type as the key.

**Review by SupunS — APPROVED** [review]

File: migrations/entitlements/migration.go [hunk 3]
```diff
@@ -20,21 +20,40 @@ package entitlements
...
 )
 
+type StaticTypeCache struct {
+	m sync.Map
+}
+
 type EntitlementsMigration struct {
-	Interpreter *interpreter.Interpreter
+	Interpreter       *interpreter.Interpreter
+	migratedTypeCache *StaticTypeCache
```
[the remark this answers is not in the document]
Comment (dsainati1) [reply migrations/entitlements/migration.go:37]: I think this is mostly for type safety? This was how @turbolent suggested to do it, but I suppose I'm not exactly sure why

**Comment by dsainati1** [comment]
> given that the entitlements migration runs after the static type migration, I guess it should be safe to use the type ID as the cache key

Yeah, this assumption was why the dependency between this and the optional typeID PR existed.

**Review by turbolent — APPROVED** [review]
👍

**Review by SupunS — APPROVED** [review]

**Pull request merged**